source/extensions/filters/http/csrf/csrf_filter.h

```diff
@@ -0,0 +1,122 @@
+  COUNTER(request_valid)        \
+// clang-format on
+
+/**
+ * Struct definition for CSRF stats. @see stats_macros.h
+ */
+struct CsrfStats {
+  ALL_CSRF_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+/**
+ * Configuration for CSRF policy.
+ */
+class CsrfPolicy : public Router::RouteSpecificFilterConfig {
+public:
+  CsrfPolicy(const envoy::config::filter::http::csrf::v2::CsrfPolicy& policy,
+             Runtime::Loader& runtime) : policy_(policy), runtime_(runtime) {}
+
+  bool enabled() const {
+    const envoy::api::v2::core::RuntimeFractionalPercent& filter_enabled = policy_.filter_enabled();
+    return runtime_.snapshot().featureEnabled(filter_enabled.runtime_key(),
+                                              filter_enabled.default_value());
+  }
+
+  bool shadowEnabled() const {
+    if (!policy_.has_shadow_enabled()) {
+      return false;
+    }
+    const envoy::api::v2::core::RuntimeFractionalPercent& shadow_enabled = policy_.shadow_enabled();
+    return runtime_.snapshot().featureEnabled(shadow_enabled.runtime_key(),
+                                              shadow_enabled.default_value());
+  }
+
+private:
+  const envoy::config::filter::http::csrf::v2::CsrfPolicy policy_;
+  Runtime::Loader& runtime_;
+};
+
+/**
+ * Configuration for the CSRF filter.
+ */
+class CsrfFilterConfig {
+public:
+  CsrfFilterConfig(const envoy::config::filter::http::csrf::v2::CsrfPolicy& policy,
+                   const std::string& stats_prefix, Stats::Scope& scope,
+                   Runtime::Loader& runtime);
+
+  CsrfStats& stats() { return stats_; }
+  const CsrfPolicy* policy() { return &policy_; }
+
+private:
+  static CsrfStats generateStats(const std::string& prefix, Stats::Scope& scope) {
+    const std::string final_prefix = prefix + "csrf.";
+    return CsrfStats{ALL_CSRF_STATS(POOL_COUNTER_PREFIX(scope, final_prefix))};
+  }
+  static const CsrfPolicy generatePolicy(const envoy::config::filter::http::csrf::v2::CsrfPolicy& policy,
+                                         Runtime::Loader& runtime) {
+    return CsrfPolicy(policy, runtime);
+  }
+
+  CsrfStats stats_;
+  const CsrfPolicy policy_;
+};
+typedef std::shared_ptr<CsrfFilterConfig> CsrfFilterConfigSharedPtr;
+
+class CsrfFilter : public Http::StreamDecoderFilter {
+public:
+  CsrfFilter(CsrfFilterConfigSharedPtr config);
+
+  // Http::StreamFilterBase
+  void onDestroy() override {}
+
+  // Http::StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::HeaderMap& headers, bool end_stream) override;
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool) override {
+    return Http::FilterDataStatus::Continue;
+  };
+  Http::FilterTrailersStatus decodeTrailers(Http::HeaderMap&) override {
+    return Http::FilterTrailersStatus::Continue;
+  };
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) override {
+    callbacks_ = &callbacks;
+  };
+
+private:
+  bool modifyMethod(const Http::HeaderMap& headers);
```

Comment:
Do these four methods need to be private? Seems they can be in an anonymous namespace in .cc file only.